lib/p2psocket.js

```diff
@@ -77,93 +101,200 @@ P2PSocket.prototype.parseInv = function(payload) {
     i += 4;
-    
-    content.vin.push(tx);      
+
+    if ( tx.vout === tx.sequence && tx.txid.match(/^0+$/) ) {
+      tx = {
+        coinbase: tx.scriptSig.hex,
+        sequence: tx.sequence,
+      }
+    }
+
+    content.vin.push(tx);
   }
 
   var cout = payload.readVarInt(i);
   i += cout.size;
-  
+
   for ( n=0; n < cout.value; n++ ) {
-  
+
     var tx = {
-      value: payload.readInt64(i),
+      value: payload.readInt64(i) / this._divisor,
     };
-    
+
     i += 8;
-    
+
     var pkl = payload.readVarInt(i);
     i += pkl.size;
-    
-    tx.scriptPubKey = payload.slice(i, i+pkl.value);
+
+    var scriptPubKey = payload.slice(i, i+pkl.value);
+    tx.scriptPubKey = {
+      hex: scriptPubKey.toString('hex'),
+      asm: scriptToAsm(scriptPubKey),
+    }
+
+    if ( scriptPubKey.length === 25 && scriptPubKey[0] === opcodes.OP_DUP &&
+         scriptPubKey[1] === opcodes.OP_HASH160 && scriptPubKey[2] === 20 &&
+         scriptPubKey[23] === opcodes.OP_EQUALVERIFY &&
+         scriptPubKey[24] === opcodes.OP_CHECKSIG
+     ) {
+      tx.scriptPubKey.reqSigs = 1;
+      tx.scriptPubKey.type = 'pubkeyhash';
+
+      var hash = scriptPubKey.slice( 3, 23 );
+
+      var addrBuf = new Buffer( hash.length + 1 );
+      addrBuf[0] = this._p2pkh_byte;
+      hash.copy( addrBuf, 1, 0 );
+
+      var sha256_pass1 = crypto.createHash('sha256').update(addrBuf).digest();
+      var sha256 = crypto.createHash('sha256').update(sha256_pass1).digest();
+
+      tx.scriptPubKey.addresses = [
+        bs58.encode( Buffer.concat( [addrBuf, sha256.slice(0,4)], addrBuf.length + 4 ) )
+      ];
+    }
+
     i += pkl.value;
-    
+
     tx.n = n;
-    
-    content.vout.push(tx);      
+
+    content.vout.push(tx);
   }
 
+  content.lock_time =  payload.readUInt32LE(i);
+
+  var raw = payload.slice(0,i+4);
+
+  var sha256_pass1 = crypto.createHash('sha256').update(raw).digest();
+  content.txid = inPlaceReverse(crypto.createHash('sha256').update(sha256_pass1).digest()).toString('hex');
+
+  content.hex = raw.toString('hex');
+
   return content;
 }
 
 P2PSocket.prototype.parse = function(method, payload) {
   switch( method ) {
-    case 'version': return this.parseVersion(payload);
-    case 'inv':     return this.parseInv(payload);
-    case 'tx':      return this.parseTx(payload);
-    default:        return null;
+    case 'version':  return this.parseVersion(payload);
+    case 'getdata':
+    case 'inv':
+    case 'notfound': return this.parseInv(payload);
+    case 'block':    return this.parseBlock(payload);
+    case 'tx':       return this.parseTx(payload);
+    default:         return null;
   }
 }
 
 P2PSocket.prototype.receive = function(data) {
   var buffer = new Buffer(data);
+  var headerSize = 24;
 
   var message = {
     raw: buffer
   };
 
   var i = 0;
-  message.magic  = buffer.slice(0,this.magic.length);
-  i += this.magic.length;
+  message.magic  = buffer.slice(0,this._magic.length);
```

Comment:
Good question. The user of the library could certainly do that in the callback function but it might make sense to do that here and maybe send an invalid packet callback. I would call that next version, though.